src/ripple/shamap/README.md

```diff
@@ -26,164 +45,322 @@ There are two different ways of building and using a SHAMap:
+already exist in the `TreeNodeCache`.
+
+The calls to `canonicalize()` make sure that if the resulting node is already in
+the SHAMap, node `TreeNodeCache` or database, then we don't create duplicates.
+
+On retrieving existing nodes, the copy in the database has top priority.  If
+there's a duplicate in the `TreeNodeCache`, it is replaced by the copy in the
+database. Or if there is a duplicate found between the `TreeNodeCache` and the
+SHAMap, the copy in the `TreeNodeCache` prevails and replaces that in the
+SHAMap.
+
+On inserting new nodes (such as new transactions or account states), the
+opposite priority holds.  The SHAMap dominates copies in the `TreeNodeCache` and
+database (if found), and the `TreeNodeCache` dominates any copies that may be
+found in the database.
+
+By using `canonicalize()` we manage a thread race condition where two different
+threads might both recognize the lack of a SHAMapTreeNode at the same time
+(during a fetch).  If they both attempt to insert the node into the SHAMap, then
+`canonicalize` makes sure that the first node in wins and the slower thread
+receives back a pointer to the node inserted by the faster thread.
+
+Conversely, two different trees can have the exact same node (the same hash
+value) with two different IDs.  If the older trie `TreeNodeCache`s the node, and
+then the node is inserted into a new trie (e.g. a transaction that didn't make
+the last ledger), then that transaction gets a new id and dominates the
+transaction node in the `TreeNodeCache`.
+
+The `TreeNodeCache` is not subject to the rule that any node must be resident
+forever.  So it's okay to replace the old node with the new node.
+
+## TreeNodeCache ##
+
+The `TreeNodeCache` is a `std::unordered_map` keyed on the hash of the
+SHAMap node.  The stored type consists of `shared_ptr<SHAMapAbstractNode>`,
+`weak_ptr<SHAMapAbstractNode>`, and a time point indicating the most recent
+access of this node in the cache.  The time point is based on
+`std::chrono::steady_clock`.
+
+The container uses a cryptographically secure hash that is randomly seeded.
+
+The `TreeNodeCache` also carries with it various data used for statistics
+and logging, and a target age for the contained nodes.  When the target age
+for a node is exceeded, and there are no more references to the node, the
+node is removed from the `TreeNodeCache`.
+
+## FullBelowCache ##
+
+This cache remembers which trie keys have all of their children resident in a
+SHAMap.  This optimizes the process of acquiring a complete trie.  This is used
+when creating the missing nodes list.  Missing nodes are those nodes that a
+SHAMap refers to but that are not stored in the local database.
+
+As a depth-first walk of a SHAMap is performed, if an inner node answers true to
+`isFullBelow()` then it is known that none of this node's children are missing
+nodes, and thus that subtree does not need to be walked.  These nodes are stored
```

Comment:
Why don't we need to walk a subtree where all children are present? Maybe an example would be useful

Reply:
The fullbelow cache is only used for the missing node list.  It exists only to optimize the "get missing nodes" operations.  When all children are present, none of them are missing.